File: src/main/java/com/jake/kotlinjava/todo/service/TodoService.java
package com.jake.kotlinjava.todo.service;

import com.jake.kotlinjava.todo.controller.model.TodoRequest;
import com.jake.kotlinjava.todo.domain.Todo;
import com.jake.kotlinjava.todo.domain.TodoRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Sort;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.Assert;
import org.springframework.web.server.ResponseStatusException;

import java.time.LocalDateTime;
import java.util.List;

@Service
public class TodoService {

	private final TodoRepository todoRepo;

	public TodoService(TodoRepository todoRepo) {
		this.todoRepo = todoRepo;
	}

	@Transactional(readOnly = true)
	public List<Todo> findAll() {
		return todoRepo.findAll(Sort.by(Sort.Direction.DESC, "id"));
	}

	@Transactional(readOnly = true)
	public Todo findById(Long id) {
		return todoRepo.findById(id)
				.orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND));
	}

	@Transactional
	public Todo create(TodoRequest request) {
		Assert.notNull(request, "TodoRequest is null");

		Todo todo = Todo.builder()
				.title(request.getTitle())
				.description(request.getDescription())
				.done(false)
				.createdAt(LocalDateTime.now())
				.build();

		return todoRepo.save(todo);
	}

	@Transactional
	public Todo update(Long id, TodoRequest request) {
		Assert.notNull(request, "TodoRequest is null");

		Todo todo = findById(id);
		todo.update(request.getTitle(),
					request.getDescription(),
					request.getDone());

		return todoRepo.save(todo);
	}

	@Transactional
	public void delete(Long id) {
		todoRepo.deleteById(id);
	}
}

File: src/main/java/com/jake/kotlinjava/todo/service/TodoService.kt
package com.jake.kotlinjava.todo.service

import com.jake.kotlinjava.todo.controller.model.TodoRequest
import com.jake.kotlinjava.todo.domain.Todo
import com.jake.kotlinjava.todo.domain.TodoRepository
import org.springframework.data.domain.Sort
import org.springframework.data.repository.findByIdOrNull
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.server.ResponseStatusException
import java.time.LocalDateTime

@Service
class TodoService(
    private val todoRepo: TodoRepository
) {

    @Transactional(readOnly = true)
    fun findAll(): List<Todo> =
        todoRepo.findAll(Sort.by(Sort.Direction.DESC, "id"))

    @Transactional(readOnly = true)
    fun findById(id: Long): Todo =
        todoRepo.findByIdOrNull(id)
            ?: throw ResponseStatusException(HttpStatus.NOT_FOUND)

    @Transactional
    fun create(request: TodoRequest?): Todo {
        checkNotNull(request) {
            "TodoRequest is null"
        }

        val todo = Todo(
            title = request.title,
            description = request.description,
            done = request.done,
            createdAt = LocalDateTime.now()
        )

        return todoRepo.save(todo)
    }

    @Transactional
    fun update(id: Long, request: TodoRequest?): Todo {
        checkNotNull(request) {
            "TodoRequest is null"
        }

        return findById(id).let {
            it.update(request.title, request.description, request.done)
            todoRepo.save(it)
        }
    }

    @Transactional
    fun delete(id: Long) = todoRepo.deleteById(id)
}